Add preset, custom_preset and custom_fan_mode support to climate

Lumpusz opened the pull request:
## Description:
The climate schema on the HA side has changed. This PR is to incorporate those changes into esphome and to add support for presets, custom_presets and custom_fan_modes to climate. There is also CLIMATE_HEAT_COOL mode which was missing from esphome. Thi PR is part of a bundle, as aioesphomeapi as well as HA esphome component needs an update.
Other related PRs:
- https://github.com/esphome/aioesphomeapi/pull/24
- https://github.com/home-assistant/core/pull/48116

Please note, that there was initial climate preset support with the away bool flag. I left it there for now, to be more backward compatible, but once the preset support is merged in component owners could migrate to it, and get rid of using the away flag, so it could be removed entirely.

**Related issue (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>
https://github.com/esphome/esphome-docs/pull/1063
## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

Lumpusz committed:
Add preset, custom_preset and custom_fan_mode support to climate

Fix lint, other minor issues

Added missing functions

Lumpusz committed:
Add preset, custom preset support for midea climate

Lumpusz committed:
Fix lint

Lumpusz committed:
Fix test

dudanov commented on esphome/components/midea_ac/midea_frame.cpp:
```diff
@@ -112,6 +178,25 @@ void PropertiesFrame::set_fan_mode(climate::ClimateFanMode mode) {
+  }
+}
+
+void PropertiesFrame::set_custom_fan_mode(std::string mode) {
```
Prefer passing objects by reference (without copy), and where the object does not change, use constant references.

dudanov commented on esphome/components/climate/climate_traits.cpp:
```diff
@@ -119,6 +119,69 @@ bool ClimateTraits::get_supports_fan_modes() const {
          this->supports_fan_mode_low_ || this->supports_fan_mode_medium_ || this->supports_fan_mode_high_ ||
          this->supports_fan_mode_middle_ || this->supports_fan_mode_focus_ || this->supports_fan_mode_diffuse_;
 }
+void ClimateTraits::set_supported_custom_fan_modes(std::vector<std::string> supported_custom_fan_modes) {
```
Be more attentive to objects. You don't use references at all. Here, in general, the object is copied twice. Pass a constant reference to the object, the assignment will call the copy constructor once, not two. Then, I haven't looked at the code above yet. If the source object is not destroyed, but is in memory, then use pointers, why copy it once more. Review your code again and optimize.

Lumpusz replied:
I come from Java, with little to no C++ knowledge. I can read the code, but I am not familiar with these differences.

dudanov reviewed (changes requested):
Be more attentive to objects. You don't use references or pointers at all. Read code comments. This is C++. ;)
Review your code again and optimize.

Lumpusz commented on esphome/components/midea_ac/climate.py:
```diff
@@ -56,8 +83,16 @@ def to_code(config):
+        cg.add(var.set_custom_fan_modes(config[CONF_CUSTOM_FAN_MODES]))
+    if CONF_CUSTOM_PRESETS in config:
+        cg.add(var.set_custom_presets(config[CONF_CUSTOM_PRESETS]))
+    cg.add(var.set_custom_fan_modes(config[CONF_CUSTOM_FAN_MODES]))
```
It is, thanks

Lumpusz committed:
Fix c&p error

Lumpusz committed:
Optimize code

Lumpusz committed:
Fix lint

Lumpusz committed:
Fix missing state update

OttoWinter commented on esphome/components/api/api.proto:
```diff
@@ -669,6 +669,7 @@ enum ClimateMode {
   CLIMATE_MODE_HEAT = 3;
   CLIMATE_MODE_FAN_ONLY = 4;
   CLIMATE_MODE_DRY = 5;
+  CLIMATE_MODE_HEAT_COOL = 6;
```
I see you changed the `AUTO` enum value in HA to map to `AUTO`. That would be a semi-breaking change (I assume the HA frontend displays it differently).

To avoid a breaking change, could we do the following:

- rename `CLIMATE_MODE_AUTO` enum value to `CLIMATE_MODE_HEAT_COOL` in esphome codebase, but leave int value 1.
- map the `CLIMATE_MODE_AUTO` enum value here to int value 6

with protobuf, only the int value is transported over the wire, so old esphome clients would be unaffected, and thus this wouldn't be a breaking change

Lumpusz replied:
I see why you would want to do that, but on the HA side it is the other way around: See here: https://github.com/esphome/aioesphomeapi/pull/24/files#diff-14008b45e808f05655e27655b1487f3255d688a74f39dd511fd5ef77d611589dR672
So if we do what you suggest, we will map `HEAT_COOL` to `AUTO` and vice-versa, which is very confusing. Do you really want to do that?

OttoWinter replied:
Ok my previous message was not that clear. No I don't want that confusing naming.

What I suggested could be summarized as: do a project-wide Find&Replace of all old occurences of `CLIMATE_MODE_AUTO` and replace it with `CLIMATE_MODE_HEAT_COOL` (including all components, ha/aioesphomeapi codebases). But do not change the associated enum values.

Then, all esphome integrations use the same name as HA for referring to climate modes, and all old clients connected to HA still show up as "heat cool" in the frontend. This is because the protobuf wire format stays the same (value 0 refers to heat/cool in the frontend), only the old esphome integrations are updated to use the new naming.

Lumpusz replied:
> What I suggested could be summarized as: do a project-wide Find&Replace of all old occurences of CLIMATE_MODE_AUTO and replace it with CLIMATE_MODE_HEAT_COOL (including all components, ha/aioesphomeapi codebases). But do not change the associated enum values.

These enums are part of the proto file only, in other places they are referred to by different names.

> (value 0 refers to heat/cool in the frontend)

0 is `NONE`

I am sorry, but you completely lost me on this.  I have been trying to wrap my head around what you said for at least an hour, and I still don't see how it would help. Probably my understanding of this is not as deep as yours. 

If we put `HEAT_COOL` as 1, and `AUTO` as 6 in both proto files, wouldn't that solve the issue?

Lumpusz replied:
I think I got it now, you want `HEAT_COOL` to stay on enum value 1, so until only HA is upgraded they are unaffected. I pushed the commits for these. Please re-review.

jesserockz replied:
> What I suggested could be summarized as: do a project-wide Find&Replace of all old occurences of CLIMATE_MODE_AUTO and replace it with CLIMATE_MODE_HEAT_COOL (including all components, ha/aioesphomeapi codebases). But do not change the associated enum values.

This was missed and is now causing all sorts of issues with the release

Lumpusz replied:
Apologies to anyone who this PR has caused inconveniences.

Lumpusz committed:
Fix AUTO to HEAT_COOL

Lumpusz committed:
Fix lint

Lumpusz committed:
Fix enum value

Lumpusz commented:
@OttoWinter @dudanov can you do another round of review

Lumpusz commented:
@glmnet @jesserockz What is still pending for this PR to be merged in? Original reviewers have not shown activity for more than 2 weeks. Happy to incorporate feedback and changes to get this rolling. Thanks.

OttoWinter reviewed (commented):
Hi, just had another look at the PR. I'm wondering now if we really should add support for "custom" fan modes/presets etc. This custom logic introduces quite a bit of complexity (any code that works with fan modes will both have to deal with custom settings and the builtin ones).

Are there enough custom fan modes/presets that we can't anticipate to warrant adding this complexity? I mean we can add new enum values for freeze protection etc, as that adds little complexity.

Obviously, adding new enum value is some work (PRs in esphome, aioesphomeapi, and HA core), but we can add all commonly used ones in this PR.

Lumpusz commented:
> Hi, just had another look at the PR. I'm wondering now if we really should add support for "custom" fan modes/presets etc. This custom logic introduces quite a bit of complexity (any code that works with fan modes will both have to deal with custom settings and the builtin ones).
> 
> Are there enough custom fan modes/presets that we can't anticipate to warrant adding this complexity? I mean we can add new enum values for freeze protection etc, as that adds little complexity.
> 
> Obviously, adding new enum value is some work (PRs in esphome, aioesphomeapi, and HA core), but we can add all commonly used ones in this PR.

This change reflects the schema of HA. Essentially it works with the enums HA works with. Wherever HA mentions support for custom values, ENUMs should be dynamically extensible. This is what I was trying to do. I personally don't agree with limiting ourselves, and diverging from the HA schema, as later on we might regret it. The goal here is to support the future development of custom values of vendor specific devices. One vendor calls boost 'Boost', others call it 'Turbo'. I understand that there are dedicated naming for some in HA, but anything more than that standard should be as close to the vendor naming convention as possible. If we introduce further enums, it will be much more effort to add yet another one later on, if there is a new device and it lacks the ones that are already there. This way the future change will be local to esphome, without the need to update HA core, and aioesphomeapi.

Lumpusz commented on esphome/components/climate/climate.h:
```diff
@@ -168,11 +190,20 @@ class Climate : public Nameable {
 
   /// The active fan mode of the climate device.
-  ClimateFanMode fan_mode;
+  optional<ClimateFanMode> fan_mode;
```
optional.has_value() drives the logic. The best would be to make it a union, but for std::string it is a no go.

glmnet replied:
couldn't `custom_fan_mode.has_value()` be checked instead?
For existing code `fan_mode` can still work as is

Lumpusz replied:
> couldn't `custom_fan_mode.has_value()` be checked instead?
> For existing code `fan_mode` can still work as is

The reason I did not do it, is because these two are mutually exclusive, and it is legit for fan_mode to not have a value in case custom_fan_mode has. Dealing with optional, which was designed for these kind of scenarios, seems to be much cleaner to me.

Lumpusz commented:
@OttoWinter what is still missing for the merge?

glmnet commented:
Agree with @Lumpusz this API is always getting a PR to add something missing that HA does. Making it 1:1 will give less space to have to adapt stuff later.
Also, since `CLIMATE_MODE_AUTO` value changed from 1 to 6, now HA would understand most climate devices are in AUTO instead of HEAT_COOL, so all occurrences of `CLIMATE_MODE_AUTO` in components should be changed to `CLIMATE_MODE_HEAT_COOL`

Lumpusz commented:
> Agree with @Lumpusz this API is always getting a PR to add something missing that HA does. Making it 1:1 will give less space to have to adapt stuff later.
> Also, since `CLIMATE_MODE_AUTO` value changed from 1 to 6, now HA would understand most climate devices are in AUTO instead of HEAT_COOL, so all occurrences of `CLIMATE_MODE_AUTO` in components should be changed to `CLIMATE_MODE_HEAT_COOL`

It seems like this meant to be a comment on the PR opened for HA core, if not then I got the wrong end of the stick.

Lumpusz commented:
Can this PR be merged? There are other related PRs as well. What is going to be the merge order? When can we expect the functionality to be available across the board?

Lumpusz commented:
@glmnet @OttoWinter @jesserockz can we merge this PR, along with its dependencies?

jesserockz commented on esphome/components/climate/__init__.py:
```diff
@@ -120,6 +138,9 @@ def register_climate(var, config):
         cv.Optional(CONF_TARGET_TEMPERATURE_HIGH): cv.templatable(cv.temperature),
         cv.Optional(CONF_AWAY): cv.templatable(cv.boolean),
         cv.Optional(CONF_FAN_MODE): cv.templatable(validate_climate_fan_mode),
+        cv.Optional(CONF_CUSTOM_FAN_MODE): cv.string_strict,
+        cv.Optional(CONF_PRESET): cv.templatable(validate_climate_preset),
+        cv.Optional(CONF_CUSTOM_PRESET): cv.string_strict,
```
Should `fan_mode`/`custom_fan_mode` and `preset`/`custom_preset` be validated as exclusive?

Lumpusz replied:
If you mean mutually exclusive, then I thing they shouldn't. There can be units who only support one or the other, or both. My midea unit supports a few built in presets, as well as a few custom one.

jesserockz replied:
But this is the control action, surely you cant select two modes at once? Also the cpp code resets "the other" one

Lumpusz replied:
Got it, fixed.

jesserockz commented on esphome/components/midea_ac/climate.py:
```diff
@@ -19,22 +19,49 @@
 CODEOWNERS = ["@dudanov"]
 
 CONF_BEEPER = "beeper"
+CONF_CUSTOM_FAN_MODES = "custom_fan_modes"
+CONF_CUSTOM_PRESETS = "custom_presets"
```
These probably should be in `const.py`, and the `PRESET_...` ones below

jesserockz commented on esphome/components/midea_ac/midea_frame.h:
```diff
@@ -82,16 +94,28 @@ class PropertiesFrame : public midea_dongle::BaseFrame {
+  bool get_eco_mode() const { return this->pbuf_[19] & 0x10; }
+  void set_eco_mode(bool state) { this->set_bytemask_(19, 0x80, state); };
 
   /* SLEEP MODE */
   bool get_sleep_mode() const { return this->pbuf_[20] & 0x01; }
-  void set_sleep_mode(bool state) { this->set_bytemask_(20, 0x01, state); }
+  void set_sleep_mode(bool state) { this->set_bytemask_(20, 0x01, state); };
 
   /* TURBO MODE */
-  bool get_turbo_mode() const { return this->pbuf_[20] & 0x02; }
-  void set_turbo_mode(bool state) { this->set_bytemask_(20, 0x02, state); }
+  bool get_turbo_mode() const { return this->pbuf_[18] & 0x20; }
+  void set_turbo_mode(bool state) { this->set_bytemask_(18, 0x20, state); };
+
+  /* FREEZE PROTECTION */
+  bool get_freeze_protection_mode() const { return this->pbuf_[31] & 0x80; }
+  void set_freeze_protection_mode(bool state) { this->set_bytemask_(31, 0x80, state); };
```
Some inconsistencies here with the semis after the functions?

jesserockz reviewed (changes requested):
A few things below and one above I hit the wrong button on

Lumpusz commented on esphome/components/climate/automation.h:
```diff
@@ -26,6 +29,9 @@ template<typename... Ts> class ControlAction : public Action<Ts...> {
     call.set_target_temperature_high(this->target_temperature_high_.optional_value(x...));
     call.set_away(this->away_.optional_value(x...));
     call.set_fan_mode(this->fan_mode_.optional_value(x...));
+    call.set_custom_fan_mode(this->custom_fan_mode_.optional_value(x...));
+    call.set_preset(this->preset_.optional_value(x...));
+    call.set_custom_preset(this->custom_preset_.optional_value(x...));
```
Those are not needed. Internally these try to convert it into a built in enum, if that is not possible, try to validate the value as a custom one, nad potentially set it as a custom mode if it matches any of the supported custom modes.

jesserockz replied:
You missed fixing these lines up, either changing them to use `set_fan_mode` and `set_preset`.
Could you add a test to the test yaml to call climate.control with the custom_fan_mode/custom_preset. 
It will fail with this code as is because `call` is a `ClimateCall` and does not have `set_custom_fan_mode` or `set_custom_preset` functions

Lumpusz replied:
Got you, fixed. I also updated the relevant section of the docs about climate control in the linked PR.

Lumpusz committed:
Fix inconsistent semis

Lumpusz committed:
Better object handling

Lumpusz committed:
Exclusively validate control action config

Lumpusz committed:
Move constants to their correct place

Lumpusz committed:
Add missing file

Lumpusz committed:
Add back test

jesserockz commented on tests/test4.yaml:
```diff
@@ -146,37 +146,3 @@ display:
       it.rectangle(3, 3, it.get_width()-6, it.get_height()-6, red);
     rotation: 0°
     update_interval: 16ms
-  - platform: waveshare_epaper
-    cs_pin: GPIO23
-    dc_pin: GPIO23
-    busy_pin: GPIO23
-    reset_pin: GPIO23
-    model: 2.13in-ttgo-b1
-    full_update_every: 30
-    lambda: |-
-      it.rectangle(0, 0, it.get_width(), it.get_height());
-  - platform: waveshare_epaper
-    cs_pin: GPIO23
-    dc_pin: GPIO23
-    busy_pin: GPIO23
-    reset_pin: GPIO23
-    model: 2.90in
-    full_update_every: 30
-    lambda: |-
-      it.rectangle(0, 0, it.get_width(), it.get_height());
-  - platform: waveshare_epaper
-    cs_pin: GPIO23
-    dc_pin: GPIO23
-    busy_pin: GPIO23
-    reset_pin: GPIO23
-    model: 2.90inv2
-    full_update_every: 30
-    lambda: |-
-      it.rectangle(0, 0, it.get_width(), it.get_height());
-
-external_components:
-  - source: github://esphome/esphome@dev
-    refresh: 1d
-    components: ["bh1750"]
-  - source: ../esphome/components
-    components: ["sntp"]
```
Not sure what happened here?

Lumpusz replied:
Accidental delete, I put it back.

Lumpusz committed:
Fix formatting

Lumpusz committed:
Fix old commits

Lumpusz committed:
Fix custom fan mode bug

Lumpusz commented:
@jesserockz I updated the PR, and also fixed a bug, which was casued by one of the suggested changes. Now it is all tested and works. Can we merge it in?

Lumpusz committed:
Add missing line

Lumpusz committed:
Fix climate control action

Lumpusz committed:
Fix return value

Lumpusz committed:
Fix test

Lumpusz commented:
@jesserockz thanks for the review, when can we merge this PR, what is going to be the order of merges into the upstream repos?

jesserockz reviewed (changes requested):
@Lumpusz Some core changes just merged which changed from `yield` to `await` syntax. Can you merge dev and update your changes?

Lumpusz committed:
Move to await syntax

Lumpusz commented:
> @Lumpusz Some core changes just merged which changed from `yield` to `await` syntax. Can you merge dev and update your changes?

@jesserockz Done.

Lumpusz commented:
@jesserockz @glmnet Not sure how long getting a second opinion takes, can you take a look at this PR? I am happy to resolve merge conflict if you think this is good enough to be merged in.

jesserockz committed:
Merge branch 'dev' into climate

Lumpusz committed:
Merge branch 'dev' into climate

Lumpusz commented:
Thanks, this was quite a journey. 🎉

kroimon commented:
This still requires the other two related PRs to be merged to be usable, correct?

Lumpusz commented:
> This still requires the other two related PRs to be merged to be usable, correct?

Correct, I am just happy, as updating these PRs was a continuous effort.
